### PR TITLE
Chapter 11 suggestions

### PR DESCRIPTION
**Pull request by aaupov**
- 11-1: suggestions
- 11-6: hfsort suggestions
- 11-8: BOLT hugify
- 11-10: Suggestions

### CODE REVIEW

File: chapters/11-Machine-Code-Layout-Optimizations/11-6 Function Reordering.md [hunk 1]
```diff
@@ -21,7 +21,20 @@ Similar to previous optimizations, function reordering improves the utilization
...
+profiling data [@HfSort]. Using this tool, engineers from Meta got a 2\%
+performance speedup of large distributed cloud applications like Facebook,
+Baidu, and Wikipedia. Since then, HFSort has been integrated into Meta's HHVM,
+LLVM BOLT, and LLD linker[^2], and is no longer recommended as a standalone
+tool.
+
+The algorithm itself has been superseded first by HFSort+, and most recently
+by Cache-Directed Sort (CDSort), with some improvements on large code footprint
+workloads.
+[TODO]: Add link to paper and implementation.
```
Comment (dendibakh) [inline chapters/11-Machine-Code-Layout-Optimizations/11-6 Function Reordering.md:37]: Is it this paper? https://arxiv.org/abs/1809.04676

Comment (dendibakh) [reply chapters/11-Machine-Code-Layout-Optimizations/11-6 Function Reordering.md:37]: And the implementation is here? https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/CodeLayout.cpp#L1035

File: chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md [hunk 2]
```diff
@@ -34,7 +34,15 @@ $ LD_PRELOAD=/usr/lib64/liblppreload.so clang++ a.cpp
...
-Besides from employing huge pages, standard techniques for optimizing I-cache performance can be used for improving ITLB performance. Namely, reordering functions so that hot functions are collocated better, reducing the size of hot regions via Link-Time Optimizations (LTO/IPO), using Profile-Guided Optimizations (PGO), and less aggressive inlining.
+Besides from employing huge pages, standard techniques for optimizing I-cache performance can be used for improving ITLB performance. Namely, reordering functions so that hot functions are collocated better, reducing the size of hot regions via Link-Time Optimizations (LTO/IPO), using Profile-Guided Optimizations (PGO) and BOLT, and less aggressive inlining.
+
+BOLT provides an option to automatically use huge pages for hot code based on
+profile data. The benefits of this approach is that only the small portion of
+the code is mapped to the huge pages and the number of required huge pages is
+minimized, and as a side effect, page fragmentation is reduced.
+The implementation leverages Linux Transparent Huge Pages (THP) for requesting
+huge pages at runtime.
+[TODO]: provide the option/documentation.
```
Comment (dendibakh) [inline chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md:45]: I couldn't find documentation for this. Only source code.

File: chapters/11-Machine-Code-Layout-Optimizations/11-6 Function Reordering.md (hunk 1 above)
Comment (aaupov) [reply chapters/11-Machine-Code-Layout-Optimizations/11-6 Function Reordering.md:37]: The implementation link is correct. Relevant benchmarking results were given in the diff https://reviews.llvm.org/D152834. 
But it seems like there's no paper yet.

File: chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md (hunk 2 above)
Comment (aaupov) [reply chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md:45]: Does BOLT option description suffice?
> Automatically put hot code on 2MB page(s) (hugify) at runtime.

Comment (dendibakh) [reply chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md:45]: Sort of. :) I will give a more detailed explanation of what it does. Please review afterward.

**Comment by dendibakh** [comment]
Amir, I improved text about `hfsort` and `-hugify`, please review.

File: chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md [hunk 3]
```diff
@@ -36,15 +34,8 @@ While the first method only works with explicit huge pages, the second approach
...
 
-BOLT provides an option to automatically use huge pages for hot code based on
-profile data. The benefits of this approach is that only the small portion of
-the code is mapped to the huge pages and the number of required huge pages is
-minimized, and as a side effect, page fragmentation is reduced.
-The implementation leverages Linux Transparent Huge Pages (THP) for requesting
-huge pages at runtime.
-[TODO]: provide the option/documentation.
+BOLT provides the `-hugify` option to automatically use huge pages for hot code based on profile data. When this option is used, `llvm-bolt` will inject a trampoline to the program's entry point, with contains the code to put hot code on 2MB pages at runtime. The implementation leverages Linux Transparent Huge Pages (THP). The benefits of this approach is that only the small portion of the code is mapped to the huge pages and the number of required huge pages is minimized, and as a side effect, page fragmentation is reduced.
```
Comment (aaupov) [inline chapters/11-Machine-Code-Layout-Optimizations/11-8 Reducing ITLB misses.md:37]:
```suggestion
BOLT provides the `-hugify` option to automatically use huge pages for hot code based on profile data. When this option is used, `llvm-bolt` will inject the code to put hot code on 2MB pages at runtime. The implementation leverages Linux Transparent Huge Pages (THP). The benefit of this approach is that only a small portion of the code is mapped to the huge pages and the number of required huge pages is minimized, and as a consequence, page fragmentation is reduced. 
```

**Review by aaupov — COMMENTED** [review]
Looks good overall, with one suggestion regarding hugify: no need to explain the implementation.

**Comment by dendibakh** [comment]
Thanks a lot, Amir, will go ahead and merge.

**Pull request merged**